### PR TITLE
Adds background job scheduling and execution

### LINKED ISSUE #11122: Background Jobs: Re-work to procedural design

### PR DESCRIPTION
**Pull request by cannikin**
This new package provides scheduling and processing of background jobs. We want everything needed to run a modern web application to be included in Redwood itself—you shouldn't need any third party integrations if you don't want. 

Background jobs have been sorely missed, but the time has come! (If you do want to use a third party service we have had an [integration with Inngest](https://community.redwoodjs.com/t/ship-background-jobs-crons-webhooks-and-reliable-workflows-in-record-time-with-inngest-and-redwoodjs/4866) since May of 2023!)

## Features

* Named queues: you can schedule jobs in separate named queues and have a different number of workers monitoring each one—makes it much easier to scale your background processing.
* Priority: give your jobs a priority from 1 (highest) to 100 (lowest). Workers will sort available jobs by priority, working on the most important ones first.
* Configurable delay: run your job as soon as possible (default), wait a number of seconds before running, or run at a specific time in the future.
* Auto-retries with backoff: if your job fails it will back off at the rate of attempts ** 4 for a default of 24 tries, the time between the last two attempts is a little over three days.
* Run inline: instead of scheduling to run in the background, run immediately.
* Integrates with Redwood's [logger](https://docs.redwoodjs.com/docs/logger): use your existing one in `api/src/lib/logger` or create a new one just for job logging.

## Usage

Check out [the docs](https://deploy-preview-11238--redwoodjs-docs.netlify.app/docs/canary/background-jobs)!

## The Future

* More adapters (maybe the community wants to get involved?): Redis, SQS, RabbitMQ
* Studio integration: monitor the state of your outstanding jobs
* Baremetal integration: if jobs are enabled, monitor the workers
* Recurring jobs
* Lifecycle hooks: `beforePerform()`, `afterPerform()`, `afterSuccess()`, `afterFailure()`

## TODO

* [x] Update docs

Thanks to @Josh-Walker-GM for the help in the recent refactor!

Closes #11122

### CODE REVIEW

**Comment by ladderschool** [comment]
This is the bomb. Bravo! Looking forward to using this, since my use case can be satisfied with RedwoodJob instead of RabbitMQ or BullMQ, which are other technologies I was researching for my use case. ❤️ 🚀

File: docs/docs/background-jobs.md [hunk 1]
```diff
@@ -0,0 +1,670 @@
...
+
+A typical create-user flow could look something like this:
+
+![image](/img/background-jobs/jobs-before.png)
+
+If we want the email to be send asynchonously, we can shuttle that process off into a **background job**:
+
+![image](/img/background-jobs/jobs-after.png)
+
+The user's response is returned much quicker, and the email is sent by another process which is connected to a user's session. All of the logic around sending the email is packaged up as a **job** and a **job worker** is responsible for executing it.
```
Comment (Josh-Walker-GM) [inline docs/docs/background-jobs.md:13]: > which is connected to a user's session

Makes me think of auth sessions and cookies. Might just be me though.

File: docs/docs/background-jobs.md [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+NODE_ENV=development
+```
+
+If `NODE_ENV` is _not_ set then the job runner will assume you're running a production-like environment and get jobs from `api/dist/jobs`. If you try to start a job worker and see this error:
+
+```
+JobsLibNotFoundError: api/src/lib/jobs.ts not found. Run `yarn rw setup jobs` to create this file and configure background jobs
+```
+
+Then `NODE_ENV` is probably not set!
```
Comment (Josh-Walker-GM) [inline docs/docs/background-jobs.md:33]: We don't need this anymore right?

File: packages/cli/src/commands/setup/jobs/jobsHandler.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+          ${c.success('\nBackground jobs configured!\n')}
+
+          ${!modelExists ? 'Migrate your database to finish setting up jobs:\n' : ''}
+          ${!modelExists ? c.warning('\n\u00A0\u00A0yarn rw prisma migrate dev\n') : ''}
+
+          Generate jobs with: ${c.warning('yarn rw g job <name>')}
+          Execute jobs with:  ${c.warning('yarn rw jobs work\n')}
+
+          Check out the docs for more info:
+          ${c.link('https://docs.redwoodjs.com/docs/background-jobs')}
```
Comment (Josh-Walker-GM) [inline packages/cli/src/commands/setup/jobs/jobsHandler.js:122]: This link colour is only just visible on a terminal with a white background.

File: packages/cli/src/commands/setup/jobs/templates/jobs.ts.template [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  adapters: {
+    prisma: new PrismaAdapter({ db, logger }),
+  },
+  queues: ['default'] as const,
+  logger,
+  workers: [
+    {
+      adapter: 'prisma',
+      logger,
+      queue: '*',
```
Comment (Josh-Walker-GM) [inline packages/cli/src/commands/setup/jobs/templates/jobs.ts.template:20]: Let's add a comment here to say what '*' means if you think that's sensible? Something like:
```
queue: '*', // Listens on queues
```

File: docs/docs/background-jobs.md [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+  maxAttempts: 24,
+  maxRuntime: 14_400,
+  sleepDelay: 5,
+  deleteFailedJobs: false,
+}
+
+RedwoodJob.config({ adapter, logger })
+
+export const jobs = {}
+```
```
Comment (Josh-Walker-GM) [inline docs/docs/background-jobs.md:84]: No longer matches the template after the recent refactoring.

File: docs/docs/background-jobs.md [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+
+We'll go into more detail on this file later (see [RedwoodJob (Global) Configuration](#redwoodjob-global-configuration)), but what's there now is fine to get started creating a job.
+
+### Creating a Job
```
Comment (Josh-Walker-GM) [inline docs/docs/background-jobs.md:88]: I'll leave reviewing the rest of the docs for the moment until you've had a chance to update them.

File: packages/jobs/src/adapters/BaseAdapter/__tests__/BaseAdapter.test.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import { describe, expect, vi, it } from 'vitest'
+
+import { BaseAdapter } from '../BaseAdapter'
+import type { BaseAdapterOptions } from '../BaseAdapter'
+
+const mockLogger = {
```
Comment (Josh-Walker-GM) [inline packages/jobs/src/adapters/BaseAdapter/__tests__/BaseAdapter.test.ts:6]: You did define a mock logger in a little mocks file if you want to use that here?

File: packages/jobs/src/adapters/PrismaAdapter/__tests__/PrismaAdapter.test.ts [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+
+import { DEFAULT_MODEL_NAME } from '../../../consts'
+import * as errors from '../errors'
+import { PrismaAdapter } from '../PrismaAdapter'
+
+vi.useFakeTimers().setSystemTime(new Date('2024-01-01'))
+
+let mockDb: PrismaClient
+
+const mockLogger = {
```
Comment (Josh-Walker-GM) [inline packages/jobs/src/adapters/PrismaAdapter/__tests__/PrismaAdapter.test.ts:12]: Again there was the mock logger in the helper file.

File: packages/jobs/src/bins/__tests__/rw-jobs-worker.test.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import { describe, expect, it } from 'vitest'
+
+// import * as worker from '../worker'
+
+describe('worker', () => {
```
Comment (Josh-Walker-GM) [inline packages/jobs/src/bins/__tests__/rw-jobs-worker.test.js:5]: Do we want to have unit tests here? I'm happy with the answer being no if this isn't a sensible place for unit testing and e2e style testing would be better?

File: packages/jobs/src/bins/__tests__/rw-jobs.test.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import { describe, expect, it } from 'vitest'
+
+// import * as runner from '../runner'
+
+describe('runner', () => {
```
Comment (Josh-Walker-GM) [inline packages/jobs/src/bins/__tests__/rw-jobs.test.js:5]: Again did we want to fill this out or forget it?

**Review by Josh-Walker-GM — COMMENTED** [review]
Nitpicking style comments. I worked with you on some of the refactoring here so the code looks reasonably familiar to me and it all looks good.

My suggestion would be:
1. We get the docs updated and looking good.
2. Merge this into main and have canary published/put it in the v8 rc
3. We get on a call with someone on the team who didn't work on this and just let them loose with it. See what they get stuck on or if they find any bugs.

But overall this looks good and with the docs update I'd be happy to merge.

File: docs/docs/background-jobs.md (hunk 1 above)
Comment (cannikin) [reply docs/docs/background-jobs.md]: Yeah not sure why I included that, I just reworded it.

File: docs/docs/background-jobs.md (hunk 2 above)
Comment (cannikin) [reply docs/docs/background-jobs.md]: Correct, removed!

File: packages/cli/src/commands/setup/jobs/jobsHandler.js (hunk 3 above)
Comment (cannikin) [reply packages/cli/src/commands/setup/jobs/jobsHandler.js:122]: Yeah we really need to update that `link` color for everyone. I like the fact that we have a separate `colors` setting for it (like underlining it, perhaps) but the color choice needs an update! Or just remove the color and only provide the underline.

File: docs/docs/background-jobs.md (hunk 5 above)
Comment (cannikin) [reply docs/docs/background-jobs.md]: Updated!

**Review by Josh-Walker-GM — APPROVED** [review]

File: packages/jobs/src/core/Worker.ts [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+  processName: string
+  queues: string[]
+  // optional
+  logger?: BasicLogger
+  clear?: boolean
+  maxAttempts?: number
+  maxRuntime?: number
+  deleteSuccessfulJobs?: boolean
+  deleteFailedJobs?: boolean
+  sleepDelay?: number
```
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/jobs/src/core/Worker.ts:31]: Yes you can, and you get `NaN`. I've addressed this in my follow-up PR

File: packages/jobs/src/core/JobManager.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+      return scheduler.schedule({ job, jobArgs, jobOptions })
+    }
+  }
+
+  createJob<TArgs extends unknown[]>(
+    jobDefinition: JobDefinition<TQueues, TArgs>,
+  ): Job<TQueues, TArgs> {
+    // The cast is necessary because the JobDefinition type lacks the `name` and
+    // `path` properties that are required by the Job type. These properties are
+    // added to the job at build time by a plugin in the build process.
```
[the remark this answers is not in the document]
Comment (Josh-Walker-GM) [reply packages/jobs/src/core/JobManager.ts:60]: Dev for the API side is just a watcher that reruns build after any changes. So if you're running the dev server you're constantly rerunning a full api build and then as a result you're getting these added for you. 

Rob mentioned on the docs that people should be careful if they're making changes and don't run either the dev server or build command. It's unlikely people would be in that situation but good to call it out regardless.

File: docs/docs/background-jobs.md [hunk 13]
```diff
@@ -0,0 +1,744 @@
...
+```js
+import { db } from 'src/lib/db'
+import { mailer } from 'src/lib/mailer'
+import { jobs } from 'src/lib/jobs'
+
+export const SendWelcomeEmailJob = jobs.createJob({
+  queue: 'default',
+  perform: async (userId) => {
+    const user = await db.user.findUnique({ where: { id: userId } })
+    mailer.send(WelcomeEmail({ user }), {
```
[the remark this answers is not in the document]
Comment (Josh-Walker-GM) [reply docs/docs/background-jobs.md:219]: Yes

[the remark this answers is not in the document]
Comment (Tobbe) [reply docs/docs/background-jobs.md:219]: Thanks. I updated this in my follow-up PR

**Pull request merged**